File: PythonLeetcode/LeetCodeE/386_LexicographicalNumbers.py


# 386. Lexicographical Numbers

# Given an integer n, return 1 - n in lexicographical order.

# For example, given 13, return: [1,10,11,12,13,2,3,4,5,6,7,8,9].

# Please optimize your algorithm to use less time and space. The input size may be as large as 5,000,000.

class lexicalOrder:
    def doit(self, n):
        """
        :type n: int
        :rtype: List[int]
        """

        def search(base, maxV):
            if base >= maxV:
                return [base] if base == maxV else []

            res = []
            for i in range(10):
                if base + i <= maxV:
                    res += [base+i] + search( (base+i) * 10, maxV)

            return res                                       

        res = []
        for i in range(1, 10):
            if i <= n:    
                res += [i] + search(i * 10, n)

        return res


    def doit1(self, n):
        """
        :type n: int
        :rtype: List[int]
        """ 
        x = 1
        res = []
        for _ in range(n):
        
            res.append(x)

            if x * 10 <= n:
                x *= 10

            else:

                if x >= n:
                    x //= 10

                x += 1

                while x % 10 == 0:
                    x //= 10

        return res


    def doit2(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        return sorted([i for i in range(1, n + 1)], key=str)


    def doit3(self, n):

        def addNumber(self, n, start, result):
            if start > 0:
                if start <= n:
                    result.append(start)
                else:
                    return result
                result = self.addNumber(n, start*10, result)
            for nextNumbers in range(start+1, min([start+10,n+1])):
                if nextNumbers <= n:
                    result.append(nextNumbers)
                    result = self.addNumber(n, nextNumbers*10, result)
                else:
                    break
            return result

        result = []
        return addNumber(n, 0, result)
    

        



if __name__=="__main__":

    
    res = lexicalOrder().doit(13)

    res = lexicalOrder().doit(100)

    res = lexicalOrder().doit(199)

    pass

File: cppLeetcode/leetcodeM/386_LexicographicalNumbers.cpp
/*
386. Lexicographical Numbers

Given an integer n, return all the numbers in the range [1, n] sorted in lexicographical order.

 

Example 1:

Input: n = 13
Output: [1,10,11,12,13,2,3,4,5,6,7,8,9]
Example 2:

Input: n = 2
Output: [1,2]
 

Constraints:

1 <= n <= 5 * 104
 

Follow up: Could you optimize your solution to use O(n) runtime and O(1) space?

*/

#include <vector>

using std::vector;


class LexicalOrder {

    /*
        386.Lexicographical-Numbers
        研究序列[1,10,11,12,13,2,3,4,5,6,7,8,9]，找出字典序的规律。

        规律1：不考虑上限，元素1后面跟什么元素？10, 100 … 也就是不断乘以10。

        规律2：如果99是上限，那么10后面的元素不能是100了，该怎么办？答案是11，也就是加1，这样个位上的数变大了。如果加1导致进位的话，虽然个位数变0，但十位上的数会变大，总之肯定字典序往后移。
        但此时得到的并不是下一个的目标，因为把其末尾的0去掉会得到字典序相对更前的数。砍掉0之后就可以重复规律1的操作了。

        规律3：如果上限是19，那么19后面的元素就不能是20了，该怎么办？答案是将19除以10，然后再重复规律2（也就是加1），也就是得到2，之后又可以重复规律1了。
    */
    vector<int> lexicalOrder(int n) 
    {
        int current=1;
        vector<int>results(n);
        
        for (int i=0; i<n; i++)
        {
            results[i]=current;
            
            if (current*10<=n)
                current=current*10;
            else
            {
                if (current+1>n) 
                    current=current/10;
                current++;
                while (current % 10==0)
                    current/=10;
            }
        }
        return results;
    }

public:
    
    vector<int> doit_greedy(int n) {
        
        vector<int> res(n);
        int current = 1;
        
        for (int i = 0; i < n; i++) {
            
            res[i] = current;
            
            if (current * 10 <= n) {
                current *= 10;
            } else {
                
                if (current + 1 > n)
                    current /= 10;
                current++;
                
                while (current % 10 == 0)
                    current /= 10;
            }
        }
        
        return res;
    }
};